src/couch_multidb_changes.erl

```diff
@@ -0,0 +1,296 @@
+% use this file except in compliance with the License. You may obtain a copy of
+% the License at
+%
+%   http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+% License for the specific language governing permissions and limitations under
+% the License.
+
+-module(couch_multidb_changes).
+-behaviour(gen_server).
+
+-export([start_link/4]).
+
+-export([init/1, handle_call/3, handle_info/2, handle_cast/2]).
+-export([code_change/3, terminate/2]).
+
+-export([changes_reader/3, changes_reader_cb/3]).
+-export([handle_db_event/3]).
+
+-include_lib("couch/include/couch_db.hrl").
+
+-define(CTX, {user_ctx, #user_ctx{roles=[<<"_admin">>, <<"_replicator">>]}}).
+
+
+-record(state, {
+    tid :: ets:tid(),
+    mod :: atom(),
+    ctx :: term(),
+    suffix :: binary(),
+    event_listener :: pid(),
+    scanner :: nil | pid(),
+    pids :: [{binary(),pid()}],
+    skip_ddocs :: boolean()
+}).
+
+
+
+% Behavior API
+
+% For each db shard with a matching suffix, report created,
+% deleted, found (discovered) and change events.
+
+-callback db_created(DbName :: binary(), Context :: term()) ->
+    Context :: term().
+
+-callback db_deleted(DbName :: binary(), Context :: term()) ->
+    Context :: term().
+
+-callback db_found(DbName :: binary(), Context :: term()) ->
+    Context :: term().
+
+-callback db_change(DbName :: binary(), Change :: term(), Context :: term()) ->
+    Context :: term().
+
+
+
+% External API
+
+
+% Opts list can contain:
+%  - `skip_ddocs` : Skip design docs
+
+-spec start_link(binary(), module(), term(), list()) ->
+    {ok, pid()} | ignore | {error, term()}.
+start_link(DbSuffix, Module, Context, Opts) when
+    is_binary(DbSuffix), is_atom(Module), is_list(Opts) ->
+    gen_server:start_link(?MODULE, [DbSuffix, Module, Context, Opts], []).
+
+
+% gen_server callbacks
+
+init([DbSuffix, Module, Context, Opts]) ->
+    process_flag(trap_exit, true),
+    Server = self(),
+    {ok, #state{
+            tid = ets:new(?MODULE, [set, protected]),
+            mod = Module,
+            ctx = Context,
+	    suffix = DbSuffix,
+	    event_listener = start_event_listener(DbSuffix),
+	    scanner = spawn_link(fun() -> scan_all_dbs(Server, DbSuffix) end),
+            pids = [],
+            skip_ddocs = proplists:is_defined(skip_ddocs, Opts)
+    }}.
+
+
+handle_call({change, DbName, Change}, _From,
+    #state{skip_ddocs=true, mod=Mod, ctx=Ctx} = State) ->
+    couch_log:info("Notifying change ~p ~p",[DbName, Change]),
+    case is_design_doc(Change) of
+        true ->
+            {reply, ok, State};
+        false ->
+            {reply, ok, State#state{ctx=Mod:db_change(DbName, Change, Ctx)}}
+    end;
+
+handle_call({change, DbName, Change}, _From, #state{mod=Mod, ctx=Ctx} = State) ->
```

Comment:
seems we don't log anything for skip_ddocs=false case. I think we should combine these clauses.